File: data-structures/queue/Queue.js
const LinkedList = require('../linked-list/LinkedList.js')

class Queue {
    /** @type {LinkedList} */
    linkedList = new LinkedList()

    /**
     * @returns {boolean}
     */
    isEmpty() {
        return !this.linkedList.head
    }

    /**
     * Read the element at the front of the queue without removing it.
     * @returns {*}
     */
    peek() {
        return this.linkedList.head?.value
    }

    /**
     * Add a new element to the end of the queue (the tail of the linked list).
     * This element will be processed after all elements ahead of it.
     * @param {*} value
     */
    enqueue(value) {
        this.linkedList.append(value)
    }

    /**
     * Remove the element at the front of the queue (the head of the linked list).
     * @returns {*}
     */
    dequeue() {
        const deletedHead = this.linkedList.deleteHead()
        return deletedHead?.value
    }
}

module.exports = Queue

File: data-structures/queue/queue.go
package queue

import (
	"container/list"
)

type Queue struct {
	list *list.List
}

func New() *Queue {
	return &Queue{
		list: list.New(),
	}
}

func (q *Queue) IsEmpty() bool {
	return q.list.Front() == nil
}

func (q *Queue) Enqueue(value interface{}) {
	q.list.PushBack(value)
}

func (q *Queue) Dequeue() interface{} {
	if el := q.list.Front(); el != nil {
		q.list.Remove(el)
		return el.Value
	}
	return nil
}

func (q *Queue) Peek() interface{} {
	if el := q.list.Front(); el != nil {
		return el.Value
	}
	return nil
}
